File: java/src/main/java/com/workshop/architecture/fitness/managing_customers/CustomerService.java
package com.workshop.architecture.fitness.managing_customers;

import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class CustomerService {

    private final CustomerRepository repository;

    public CustomerService(CustomerRepository repository) {
        this.repository = repository;
    }

    public List<CustomerResponse> findAll() {
        return repository.findAll().stream()
                .sorted(Comparator.comparing(CustomerEntity::getName))
                .map(CustomerResponse::fromEntity)
                .toList();
    }

    public CustomerResponse findById(UUID customerId) {
        return CustomerResponse.fromEntity(load(customerId));
    }

    public CustomerResponse create(CustomerUpsertRequest request) {
        CustomerEntity entity = new CustomerEntity(
                UUID.randomUUID(),
                request.name(),
                request.dateOfBirth(),
                request.emailAddress()
        );
        return CustomerResponse.fromEntity(repository.save(entity));
    }

    public CustomerResponse update(UUID customerId, CustomerUpsertRequest request) {
        CustomerEntity entity = load(customerId);
        entity.updateFrom(request);
        return CustomerResponse.fromEntity(repository.save(entity));
    }

    public void delete(UUID customerId) {
        CustomerEntity entity = load(customerId);
        repository.delete(entity);
    }

    private CustomerEntity load(UUID customerId) {
        return repository.findById(customerId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Customer %s was not found".formatted(customerId)
                ));
    }
}

File: python/src/workshop_api/fitness/layered/business/customer_service.py
from uuid import UUID

from workshop_api.fitness.layered.business.errors import NotFoundError
from workshop_api.fitness.layered.infrastructure.customer_entity import CustomerOrmModel
from workshop_api.fitness.layered.infrastructure.customer_repository import CustomerRepository
from workshop_api.fitness.layered.presentation.customer_schemas import (
    CustomerResponse,
    CustomerUpsertRequest,
)


class CustomerService:
    def __init__(self, repository: CustomerRepository) -> None:
        self.repository = repository

    def list_customers(self) -> list[CustomerResponse]:
        customers = sorted(self.repository.find_all(), key=lambda customer: customer.name)
        return [self._to_response(customer) for customer in customers]

    def get_customer(self, customer_id: UUID) -> CustomerResponse:
        return self._to_response(self._load_customer(customer_id))

    def create_customer(self, request: CustomerUpsertRequest) -> CustomerResponse:
        customer = CustomerOrmModel(
            name=request.name,
            date_of_birth=request.date_of_birth,
            email_address=str(request.email_address),
        )
        return self._to_response(self.repository.save(customer))

    def update_customer(
        self,
        customer_id: UUID,
        request: CustomerUpsertRequest,
    ) -> CustomerResponse:
        customer = self._load_customer(customer_id)
        customer.name = request.name
        customer.date_of_birth = request.date_of_birth
        customer.email_address = str(request.email_address)
        return self._to_response(self.repository.save(customer))

    def delete_customer(self, customer_id: UUID) -> None:
        customer = self._load_customer(customer_id)
        self.repository.delete(customer)

    def _load_customer(self, customer_id: UUID) -> CustomerOrmModel:
        customer = self.repository.find_by_id(customer_id)
        if customer is None:
            raise NotFoundError(f"Customer {customer_id} was not found")
        return customer

    @staticmethod
    def _to_response(customer: CustomerOrmModel) -> CustomerResponse:
        return CustomerResponse(
            id=customer.id,
            name=customer.name,
            dateOfBirth=customer.date_of_birth,
            emailAddress=customer.email_address,
        )
